### PR TITLE
Accept/reject/proposals

### PR DESCRIPTION
**Pull request by casweeney**
### Changes made:
- Created a `custom` folder and a `types.cairo` file to hold all the custom types.
- Changed the name of the `BorrowProposal` struct to `Proposal`.
- Created an enum `ProposalType` with two fields: `BORROWING` and `LENDING` to capture what a proposal if meant for.
- Added `id`, `lender`, `proposal_type`, `accepted_collateral_token`, `required_collateral_value`, `is_accepted`, `accepted_at`, `repayment_date` and `is_repaid` fields to the `Proposal` struct.
- Created new storage variables: `proposals`, `proposals_count`, `protocol_fee_address`, `spok_nft`, `next_spok_id` and `locked_collateral` to make the smart contract work as expected.
- Created constants `COLLATERAL_RATIO_NUMERATOR`, `COLLATERAL_RATIO_DENOMINATOR` and `PROTOCOL_FEE_PERCENTAGE` for accurate collateral calculation.
- Updated the `constructor()` function to set `protocol_fee_address` and `spok_nft`.
- Refactored the `create_borrow_proposal()` function and corrected logic errors: Added support for collateral token and collateral token value.
- Updated implementation factored that the borrower has 1.3x the amount of the required collateral token.
- Added a helper function `zero_address()` to easily make use of zero address in the contract.
- Updated `TransactionType` enum and added `LEND` and `BORROW` variants to track lending and borrowing transactions.
- Created 4 extra helper functions: `handle_borrower_acceptance()`, `handle_lender_acceptance()`, `mint_spoks()` and `record_transaction()`.
- The `handle_borrower_acceptance()` handles acceptance for a borrow proposal while the `handle_lender_acceptance()` handles acceptance for a lending proposal.
- `mint_spoks()` mints SPOK to lender and borrower.
- `record_transaction()` records transaction in the transaction history, to keep the code DRY.
- Updated the `deposit()` and `withdraw()` function to use the `record_transaction()` function
- Updated test cases to all work with all the above changes.

### CODE REVIEW

File: contracts/src/interfaces.cairo [hunk 1]
```diff
@@ -1,2 +1,3 @@
 pub mod ierc20;
-pub mod ipeer_protocol;
\ No newline at end of file
+pub mod ipeer_protocol;
+pub mod ierc721;
```
Comment (Dprof-in-tech) [inline contracts/src/interfaces.cairo:3]: Screenshot 2024-11-24 at 05 49 44
please fix so it builds

File: contracts/src/peer_protocol.cairo [hunk 2]
```diff
@@ -385,6 +432,108 @@ pub struct BorrowProposalCreated {
...
+            assert(lender_balance >= proposal.amount, 'insufficient lender balance');
+
+            // Transfer net amount to borrower
+            IERC20Dispatcher { contract_address: proposal.token }.transfer(proposal.borrower, net_amount);
+
+            // Transfer protocol fee to protocol fee address
+            IERC20Dispatcher { contract_address: proposal.token }.transfer(self.protocol_fee_address.read(), fee_amount);
+
+            // Mint SPOK
+            self.mint_spoks(proposal.borrower, lender);
```
Comment (Dprof-in-tech) [inline contracts/src/peer_protocol.cairo:452]: Screenshot 2024-11-24 at 05 49 44

**Review by Dprof-in-tech — CHANGES_REQUESTED** [review]

File: contracts/src/interfaces.cairo (hunk 1 above)
Comment (casweeney) [reply contracts/src/interfaces.cairo:3]: @Dprof-in-tech I don't understand why your build is failing. Everything works fine on my end.

File: contracts/src/peer_protocol.cairo (hunk 2 above)
Comment (casweeney) [reply contracts/src/peer_protocol.cairo:452]: @Dprof-in-tech I just saw the issue. `src` was added to `gitignore` making newly created files hidden.

**Review by Dprof-in-tech — APPROVED** [review]
this is well done chief. splendid

**Comment by casweeney** [comment]
> this is well done chief. splendid

Thank you.

**Pull request merged**